Python/chapter02/2.8 - Loop Detection/miguel_soln_2.8.py

```diff
@@ -183,8 +183,20 @@ def loop_detection_linear_time_const_space(ll: LinkedList):
     while slow_ptr and fast_ptr and fast_ptr.next:
         slow_ptr = slow_ptr.next
         fast_ptr = fast_ptr.next.next
-        if slow_ptr is fast_ptr or slow_ptr is fast_ptr.next:
-            return slow_ptr
+        if slow_ptr is not fast_ptr:
+            continue
+        # if we get here, then there is a cycle.
+        # advance one of fast or slow pointers
+        # and a pointer that starts in the
+        # beginning, by one until they match.
+        # they will end at the beginning of
+        # the cycle.
+        p = ll.head
+        while p and slow_ptr:
```

Comment:
given that there is a cycle, `p` and `slow_ptr` will never be `None`, right? you can instead do

```python
while p is not slow_ptr:
    p = p.next
    slow_ptr = slow_ptr.next
return p
```